ubcpi/answer_pool.py

```diff
@@ -0,0 +1,240 @@
+import random
+import persistence as sas_api
+
+
+class UnknownChooseAnswerAlgorithm(Exception):
+    pass
+
+
+def offer_answer(pool, answer, rationale, student_id, algo):
+    """
+    Answers format:
+    {
+        option1_index: {
+            'student_id': { can store algorithm specific info here },
+            ...
+        }
+        option2_index: ...
+    }
+    """
+    if algo['name'] == 'simple':
+        offer_simple(pool, answer, rationale, student_id)
+    elif algo['name'] == 'random':
+        offer_random(pool, answer, rationale, student_id)
+    else:
+        raise UnknownChooseAnswerAlgorithm()
+
+
+def offer_simple(pool, answer, rationale, student_id):
+    existing = pool.setdefault(answer, {})
+    if len(existing) >= 50:
+        student_id_to_remove = random.choice(existing.keys())
+        del existing[student_id_to_remove]
+    existing[student_id] = {}
+    pool[answer] = existing
+
+
+def offer_random(pool, answer, rationale, student_id):
+    offer_simple(pool, answer, rationale, student_id)
+
+
+def validate_seeded_answers_simple(answers, options, algo):
+    """
+    This validator checks if the answers includes all possible options
+    :param answers: the answers to be checked
+    :param options: all options that should exist in the answers
+    :return: None if everything is good. Otherwise, the missing option error message.
+    """
+    seen_options = {}
+    for answer in answers:
+        if answer:
+            key = options[answer['answer']].get('text')
+            if options[answer['answer']].get('image_url'):
+                key += options[answer['answer']].get('image_url')
+            seen_options.setdefault(key, 0)
+            seen_options[key] += 1
+
+    missing_options = []
+    index = 1
+    for option in options:
+        key = option.get('text') + option.get('image_url') if option.get('image_url') else option.get('text')
+        if seen_options.get(key, 0) == 0:
+            missing_options.append('Option ' + str(index))
+        index += 1
+
+    if missing_options:
+        return {'seed_error': 'Missing option seed(s): ' + ', '.join(missing_options)}
+
+    return None
+
+
+def validate_seeded_answers_random(answers):
+    """
+    This validator checks if there are a minimum of one answer
+    :param answers: the answers to be checked
+    :return: None if everything is good. Otherwise, the missing option error messag.
+    """
+    if len(answers) < 1:
+        return {'seed_error': 'Missing 1 option seed'}
+
+    return None
+
+
+def validate_seeded_answers(answers, options, algo):
```

Comment:
This is called when instructor setup the tool and providing seeded answers to the question. This function is trying to validate if instructor provided enough seeds for a give algorithm. e.g. we require 1 seed for each option in simple algorithm and at least 1 seed for random algorithm. Because otherwise, the first student won't be able to see the answers on the second step where he/she suppose to compare and review other students answers.